### PR TITLE
[APPS-4384] Execute postCommand after all Spryks were executed

### PR DESCRIPTION
**Pull request by kopylets**
https://spryker.atlassian.net/browse/APPS-4384

### CODE REVIEW

**Review by stereomon — CHANGES_REQUESTED** [review]

File: src/Spryk/Model/Spryk/Executor/SprykExecutor.php [hunk 1]
```diff
@@ -281,26 +287,43 @@ protected function executePreCommands(SprykDefinitionInterface $sprykDefinition,
...
+     * @return void
+     */
+    protected function storeSpryksWithPostCommands(SprykDefinitionInterface $sprykDefinition): void
+    {
+        if (!$sprykDefinition->getPostCommands()) {
+            return;
+        }
+
+        $this->spryksWithPostCommands[] = $sprykDefinition;
+    }
```
[the remark this answers is not in the document]
Comment (kopylets) [reply src/Spryk/Model/Spryk/Executor/SprykExecutor.php:325]: In this case, we will lose information about which spryk the command belongs to and we will not be able to correctly execute AbstractSprykCommand::execute() (check isRunnable and so on)

File: tests/SprykerSdkTest/Spryk/Console/SprykRunPostCommandTest.php [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+            'command' => $command->getName(),
+            SprykRunConsole::ARGUMENT_SPRYK => 'SprykWithPostCommand',
+            '--mode' => 'core',
+        ];
+
+        $currentCallIndex = 0;
+
+        $this->verifyCallIndex(
+            StructureSpryk::class,
+            'SprykerSdk\\Spryk\\Model\\Spryk\\Builder\\Structure\\StructureSpryk',
```
[the remark this answers is not in the document]
Comment (kopylets) [reply tests/SprykerSdkTest/Spryk/Console/SprykRunPostCommandTest.php:52]: How does consecutive help me check that all postСommands are run after all the spryks have been executed?

File: src/Spryk/Model/Spryk/Executor/SprykExecutor.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (stereomon) [reply src/Spryk/Model/Spryk/Executor/SprykExecutor.php:325]: I updated the example code. Now you have the command name and the definition in the cache.

File: tests/SprykerSdkTest/Spryk/Console/SprykRunPostCommandTest.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (stereomon) [reply tests/SprykerSdkTest/Spryk/Console/SprykRunPostCommandTest.php:52]: Ah ok, now I see. But you don't need to IMO. When you move the call to that method after all Spryks were executed it's fine. You can assure that the method is not moved when you add a DocBlock to that method. E.g. `For performance reasons this method must be executed after all Spryks were executed.`

We usually never test method B is called after method A. Anyways it is good to see that you thought about this. Sometimes it doesn't make sense to add such complex tests when we can ensure the same by a good explanation on top of a method. Of course, it's not an automated test but in this case, the maintainability becomes to hard compared to a comment.

File: tests/SprykerSdkTest/Spryk/Console/SprykRunPostCommandTest.php [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+        $arguments = [
+            'command' => $command->getName(),
+            SprykRunConsole::ARGUMENT_SPRYK => 'SprykWithPostCommand',
+            '--mode' => 'core',
+        ];
+
+        $currentCallIndex = 0;
+
+        $this->verifyCallIndex(
```
[the remark this answers is not in the document]
Comment (kopylets) [reply tests/SprykerSdkTest/Spryk/Console/SprykRunPostCommandTest.php]: Removed now.

File: src/Spryk/Model/Spryk/Executor/SprykExecutor.php [hunk 4]
```diff
@@ -77,9 +77,9 @@ class SprykExecutor implements SprykExecutorInterface
     protected ConditionMatcherInterface $conditionMatcher;
 
     /**
-     * @var array<\SprykerSdk\Spryk\Model\Spryk\Definition\SprykDefinitionInterface>
+     * @var array<string, \SprykerSdk\Spryk\Model\Spryk\Executor\ConditionMatcher\ConditionMatcherInterface>
```
Comment (stereomon) [inline src/Spryk/Model/Spryk/Executor/SprykExecutor.php]: The wrong interface is used here, it MUST be `SprykDefinitionInterface`

Comment (kopylets) [reply src/Spryk/Model/Spryk/Executor/SprykExecutor.php]: Already fixed in later commit

File: tests/_support/Module/SprykHelper.php [hunk 5]
```diff
@@ -329,4 +330,20 @@ protected function getContainer(): ContainerInterface
...
     }
+
+    /**
+     * @param \Codeception\Test\Unit $testClass
+     * @param string $class
+     * @param int $expectedNumberOfCalls
+     *
+     * @return void
+     */
+    public function mockCommand(Unit $testClass, string $class, int $expectedNumberOfCalls): void
```
Comment (stereomon) [inline tests/_support/Module/SprykHelper.php:341]: This method is particular to get a mock with an expected number of calls to a method. I would change it to the following

```suggestion
    public function getMockWithExpectedNumberOfMethodCalls(string $class, string, $method, int $expectedNumberOfCalls = 1): void
```
TBH I never used to pass the testCase to the Stub, what is the reason for doing it?

Comment (kopylets) [reply tests/_support/Module/SprykHelper.php]: Renamed. Without passing a testCase, fewer numbers of calls method than specified will not cause an error.

**Review by stereomon — APPROVED** [review]

**Pull request merged**